sys/net/application_layer/dns/dns.c

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright (C) 2017 Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup net_sock_dns
+ * @{
+ * @file
+ * @brief   sock DNS client implementation
+ * @author  Kaspar Schleiser <kaspar@schleiser.de>
+ * @}
+ */
+
+#include <string.h>
+#include <stdio.h>
+
+#include "net/sock/udp.h"
+#include "net/sock/dns.h"
+
+/* min domain name length is 1, so minimum record length is 7 */
+#define DNS_MIN_REPLY_LEN   (unsigned)(sizeof(sock_dns_hdr_t ) + 7)
+
+ssize_t _enc_domain_name(uint8_t *out, const char *domain_name)
+{
+    uint8_t *part_start = out;
+    uint8_t *out_pos = ++out;
+
+    char c;
+
+    while ((c = *domain_name)) {
+        if (c == '.') {
+            *part_start = (out_pos - part_start - 1);
```

Comment:
DNS encodes domain names with "<len><part><len><part>", e.g., "example.org" ends up as "\7example\2de" in the packet. I've added a note.

As for char vs uint8_t: writing the value through to the pointer does not change its binary value, right?